packages/sdk/src/search/Provider.tsx

```diff
@@ -0,0 +1,47 @@
+import React, { createContext, useEffect, useMemo } from 'react'
+import type { FC } from 'react'
+
+import { format } from './serializer'
+import { useSearchInfiniteState } from './useInfiniteSearchState'
+import { useSearchState } from './useSearchState'
+import type { State as SearchState } from './useSearchState'
+
+export interface SearchContext
+  extends ReturnType<typeof useSearchInfiniteState>,
+    ReturnType<typeof useSearchState> {
+  itemsPerPage: number
+}
+
+export const Context = createContext<SearchContext | undefined>(undefined)
+
+type Props = SearchState & {
+  onChange: (url: URL) => void
+  itemsPerPage: number
+}
+
+export const Provider: FC<Props> = ({
+  children,
+  itemsPerPage,
+  onChange,
+  ...rest
+}) => {
+  const { state, ...searchActions } = useSearchState(rest)
+  const { pages, ...infiniteActions } = useSearchInfiniteState(state.page)
+
+  useEffect(() => {
+    onChange(format(state))
+  }, [onChange, state])
+
+  const value = useMemo(
+    (): SearchContext => ({
+      state,
+      ...searchActions,
+      pages,
+      ...infiniteActions,
+      itemsPerPage,
+    }),
+    [infiniteActions, itemsPerPage, pages, searchActions, state]
+  )
```

Comment:
The idea in here is not to avoid expensive computations, but rather to avoid components that depends on this context from rendering when this component renders.

Reply:
Hmm, nice, so if you change the reference of value, react dispatch a new state from context, then the component that uses the context hook rerender?

Reply:
Yes, take a look at:  https://reactjs.org/docs/context.html#caveats